### ISSUE TITLE
Query got stuck due to tuple visibility inconsistency

### ISSUE THREAD

**Issue by pengzhout** [greenplum-db/gpdb#1275]
Reproduce step:
```
gpadmin=# begin;
BEGIN
gpadmin=# savepoint x;
SAVEPOINT
gpadmin=# create table abc (a int);
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'a' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
CREATE TABLE
gpadmin=# rollback to x;
ROLLBACK
gpadmin=# insert xcxx;
ERROR:  syntax error at or near "xcxx"
LINE 1: insert xcxx;
               ^
gpadmin=# end;
ROLLBACK
gpadmin=# begin;
BEGIN
gpadmin=# create table abc (a int);
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'a' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
<<< this statement got stuck
```

Due to codes list bellow, ShmemVariableCache->latestCompletedXid is not updated when aborting sub-transaction, and it may cause subsequent tuple visibility check incorrect.

```
RecordTransactionAbort()
{
1785:
	if (isSubXact && isQEReader)
		XidCacheRemoveRunningXids(xid, nchildren, children, latestXid);
}
```

@hlinnaka this isQEReader check was involved in commit a453004ea, aren't !isQEReader more reasonable here?

@ashwinstar @asimrp please take a look.

**Comment by hlinnaka** [comment]
On 11/03/2016 09:53 AM, Tang Pengzhou wrote:
> ```
> RecordTransactionAbort()
> {
> 1785:
> 	if (isSubXact && isQEReader)
> 		XidCacheRemoveRunningXids(xid, nchildren, children, latestXid);
> }
> ```
>
> @hlinnaka this isQEReader check was involved in commit a453004ea, aren't !isQEReader more reasonable here?

Yeah, seems so at a quick glance.

I'm a bit confused of how all this works. What does a QE reader process 
have as its top transaction ID, and in its subxid cache, in the PGPROC 
entry? Can a QE reader process acquire a (sub-)XID independently of the 
QE writer process? This are needs some more comments. More assertions to 
check those invariants would help too, also as a form of documentation.

- Heikki

**Comment by ryantang** [comment]
@Eulerizeit Does it make sense for you (on behalf of your team) to be assigned this issue?

**Comment by Eulerizeit** [comment]
We've got it.

**Comment by ashwinstar** [comment]
Fixed by a8c69af09a11ba215f662c59b1a7ef4d3610663e.
